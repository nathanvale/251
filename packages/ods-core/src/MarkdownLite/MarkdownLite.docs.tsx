/* eslint-disable react/display-name */
import React from "react";
import { ComponentDocs } from "@origin-digital/ods-types";
import { Strong, Text, TextLink } from "@origin-digital/ods-core";
import { MarkdownLite, MarkdownLiteProps } from "./MarkdownLite";

export const docs: ComponentDocs<MarkdownLiteProps> = {
  category: "Content",
  componentName: "MarkdownLite",

  description: (
    <Text>
      MarkdownLite is used to convert basic markdown into origin content
      components. <br />
      <Strong>Matches</Strong>
      <ul>
        <li>
          <Strong>Bold:</Strong> ** and __
        </li>
        <li>
          <Strong>Italic:</Strong> * and _
        </li>
        <li>
          <Strong>Links:</Strong> [Link text](http://example.com)
        </li>
      </ul>
    </Text>
  ),
  propDescriptions: {},
  migrationGuide: false,
  examples: {
    default: {
      Code: () => (
        <Text>
          <MarkdownLite>
            **Welcome** to [Origin Energy](http://originenergy.com.au)
          </MarkdownLite>
        </Text>
      ),
    },
    additional: [
      {
        label: "Mesh event implementation",
        description: (
          <Text>
            To use a mesh event in the markdown component you will need to add a
            custom linkComponent into the{" "}
            <TextLink href="/components/OriginThemeProvider">
              OriginThemeProvider
            </TextLink>{" "}
            which parses the href of the links and converts them to mesh events
          </Text>
        ),
        codeString: `
        import React from 'react';
        import { OriginThemeProvider } from "@origin-digital/ods-core";
        import { LinkComponentProps } from "@origin-digital/ods-types";
        import { parseMarkdownEvent } from "@origin-digital/event-dispatcher";

        // First create the custom link implementation
        const MeshLink = React.forwardRef<any, LinkComponentProps>((props, ref) => {
          const { href, onClick } = parseMarkdownEvent(props.href);
          return <a ref={ref} {...props} href={href} onClick={onClick} />
        });

        // Then pass it to OriginThemeProvider:
        export const App = () => (
          <OriginThemeProvider linkComponent={MeshLink}>
            ...
          </OriginThemeProvider>
        );
      `,
      },
    ],
  },
  snippets: [
    {
      label: "Basic",
      Code: () => (
        <Text>
          <MarkdownLite>
            **Bold text** and *italic text* with
            [links](http://originenergy.com.au)
          </MarkdownLite>
        </Text>
      ),
    },
  ],
};
