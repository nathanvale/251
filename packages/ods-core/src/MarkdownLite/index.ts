export { MarkdownLite } from "./MarkdownLite";
export type MarkdownLiteProps = import("./MarkdownLite").MarkdownLiteProps;
