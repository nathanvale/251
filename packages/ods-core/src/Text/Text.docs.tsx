import React from "react";
import { ComponentDocs } from "@origin-digital/ods-types";
import { Box } from "../Box/Box";
import { Stack } from "../Stack/Stack";
import { Text, TextProps } from "./Text";

export const docs: ComponentDocs<TextProps> = {
  category: "Content",
  componentName: "Text",
  description: "",
  propDescriptions: {
    variant: "The font style of the text.",
    tone:
      "Tones represent colours. The usage of colour in our design system is designed to have a strong correlation with the tone of voice being used.",
    truncate:
      "This will add an ellipsis to the end of the text if the text drops onto more than one line.",
    weight: "The font weight of the text.",
    inline:
      "Setting text to inline prevents the text node from take its parents full width.",
  },
  migrationGuide: false,
  examples: {
    default: {
      Code: () => <Text>The quick brown fox jumps over the lazy dog.</Text>,
    },
    additional: [
      {
        label: "Body Small Text",
        Code: () => (
          <Text variant="body-small">
            The quick brown fox jumps over the lazy dog.
          </Text>
        ),
      },
      {
        label: "Subtitle Text",
        Code: () => (
          <Text variant="subtitle">
            The quick brown fox jumps over the lazy dog.
          </Text>
        ),
      },
      {
        label: "Subtitle Small Text",
        Code: () => (
          <Text variant="subtitle-small">
            The quick brown fox jumps over the lazy dog.
          </Text>
        ),
      },
      {
        label: "Caption Text",
        Code: () => (
          <Text variant="caption">
            The quick brown fox jumps over the lazy dog.
          </Text>
        ),
      },
      {
        label: "Overline Text",
        Code: () => (
          <Text variant="overline-text">
            The quick brown fox jumps over the lazy dog.
          </Text>
        ),
      },
      {
        label: "Tone (positive)",
        Code: () => (
          <Text tone="positive">
            The quick brown fox jumps over the lazy dog.
          </Text>
        ),
      },
      {
        label: "Tone (critical)",
        Code: () => (
          <Text tone="critical">
            The quick brown fox jumps over the lazy dog.
          </Text>
        ),
      },
      {
        label: "Weight (medium)",
        Code: () => (
          <Text weight="medium">
            The quick brown fox jumps over the lazy dog.
          </Text>
        ),
      },
      {
        label: "Truncating long text",
        Code: () => (
          <Box style={{ width: 90 }}>
            <Text truncate>The quick brown fox jumps over the lazy dog.</Text>
          </Box>
        ),
      },
      {
        label: "Text Alignment",
        Code: () => (
          <Stack space="medium">
            <Text>The quick brown fox jumps over the lazy dog.</Text>
            <Text align="center">
              The quick brown fox jumps over the lazy dog.
            </Text>
            <Text align="right">
              The quick brown fox jumps over the lazy dog.
            </Text>
          </Stack>
        ),
      },
      {
        label: "Text Alignment (responsive)",
        Code: () => (
          <Text align={["left", "center"]}>
            The quick brown fox jumps over the lazy dog.
          </Text>
        ),
      },
      {
        label: "Inline Text",
        Code: () => (
          <Text>
            The quick brown fox{" "}
            <Text inline tone="positive">
              jumps over
            </Text>{" "}
            the lazy dog.
          </Text>
        ),
      },
    ],
  },
  snippets: [
    {
      label: "Body Text",
      Code: () => <Text>Body text</Text>,
    },
    {
      label: "Tone (positive)",
      Code: () => <Text tone="positive">Positive text</Text>,
    },
    {
      label: "Tone (critical)",
      Code: () => <Text tone="critical">Critical text</Text>,
    },
    {
      label: "Weight (medium)",
      Code: () => <Text weight="medium">Medium text</Text>,
    },
  ],
};
