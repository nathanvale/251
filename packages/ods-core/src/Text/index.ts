export { Text } from "./Text";
export { TextContextProvider, TextContext } from "./TextContextProvider";
export { useTextStyles, useTextStylesFromContext } from "./useTextStyles";

export type TextProps = import("./Text").TextProps;
export type UseTextProps = import("./TextContextProvider").UseTextProps;
export type TextContextProviderProps = import("./TextContextProvider").TextContextProviderProps;
