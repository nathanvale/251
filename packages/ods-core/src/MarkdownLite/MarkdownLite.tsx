import React, { ReactNodeArray, useContext } from "react";
import { kebabCase } from "@origin-digital/ods-helpers";
import { Emphasis } from "../Emphasis/Emphasis";
import { Strong } from "../Strong/Strong";
import { TextContext } from "../Text/TextContextProvider";
import { TextLink } from "../TextLink/TextLink";

export interface MarkdownLiteProps {
  children: string;
}

interface TagMatchConfig {
  tag: string;
  Component?: typeof Emphasis | typeof Strong;
  pattern: RegExp;
  replaceRegex: RegExp;
}

const linkRegex: TagMatchConfig = {
  // Matches: [My inline link description](https://my-inline-link-url.com)
  // Additional Matches: [My link](https://new-tab-link.com){:target="_blank"}
  tag: "a",
  pattern: /\[([^\]]*)\]\(([^)]*)\)(\{:target="(.*?)"\})?/gi,
  replaceRegex: /\[([^\]]*)\]\(([^)]*)\)(\{:target="(.*?)"\})?/,
};

const boldRegex: TagMatchConfig = {
  // Matches: __bold text__ OR **bold text**
  tag: "b",
  Component: Strong,
  pattern: /(\*\*|__)(.*?)\1/gi,
  replaceRegex: /(\*\*|__)(.*?)\1/,
};

const italicRegex: TagMatchConfig = {
  // Matches: _italic text_ OR *italic text*
  tag: "i",
  Component: Emphasis,
  pattern: /(\*|_)(.*?)\1/gi,
  replaceRegex: /(\*|_)(.*?)\1/,
};

const processNonLinkTags = (
  matchedObj: string,
  regex: TagMatchConfig,
  render: ReactNodeArray,
  startIndex: number,
  text: string
) => {
  // @ts-ignore
  const [matchOriginal, _, matchText] = matchedObj.match(regex.replaceRegex);
  const textSection = text.slice(startIndex);
  const newStartIndex = textSection.indexOf(matchOriginal);
  const currentIndex = startIndex + newStartIndex + matchOriginal.length;

  const { Component } = regex;

  render.push(textSection.slice(0, newStartIndex));

  if (matchText.length === 0) {
    render.push(matchedObj);
  } else {
    render.push(
      // @ts-ignore Component will never be undefined at this point
      <Component key={`${Component.displayName}-${startIndex}-${currentIndex}`}>
        {matchText}
      </Component>
    );
  }

  return {
    currentRender: render,
    currentIndex,
  };
};

const processLinkTags = (
  matchedObj: string,
  regex: TagMatchConfig,
  render: ReactNodeArray,
  startIndex: number,
  text: string
) => {
  // @ts-ignore
  const [linkOriginal, linkText, href, _, target] = matchedObj.match(
    regex.replaceRegex
  );

  // incase there are mutiple links which will match we want to start from the start
  // of the currentIndex not 0
  const textSection = text.slice(startIndex);
  const linkStartIndex = textSection.indexOf(linkOriginal);
  const currentIndex = startIndex + linkStartIndex + linkOriginal.length;

  render.push(textSection.slice(0, linkStartIndex));

  render.push(
    <TextLink
      key={`link-${startIndex}`}
      data-id={kebabCase(linkText)}
      href={href}
      target={target}
    >
      {linkText}
    </TextLink>
  );

  return {
    currentRender: render,
    currentIndex,
  };
};

const processEmptyRender = (
  text: string,
  regex: TagMatchConfig,
  render: ReactNodeArray
) => {
  const matchArray = text.match(regex.pattern);
  let startIndex = 0;
  if (matchArray) {
    for (const [index, matchedObj] of matchArray.entries()) {
      const processFunction =
        regex.tag === "a" ? processLinkTags : processNonLinkTags;

      const renderIndexObj = processFunction(
        matchedObj,
        regex,
        render,
        startIndex,
        text
      );

      startIndex = renderIndexObj.currentIndex;
      // If after processing there is any string left, add that back to the original array.
      if (index + 1 === matchArray.length) {
        render.push(text.slice(startIndex, text.length));
      }
    }
  }
};

const processNonEmptyRender = (
  regex: TagMatchConfig,
  render: ReactNodeArray
) => {
  /*
    Since we have already process some elements we need to
    only process the string elements
  */
  const newRender: ReactNodeArray = [];
  render.forEach((renderElement) => {
    if (typeof renderElement === "string") {
      const processedRender: ReactNodeArray = [];
      processEmptyRender(renderElement, regex, processedRender);
      if (processedRender.length) {
        newRender.push(...processedRender);
      } else {
        newRender.push(renderElement);
      }
    } else {
      // the children of our components may still need to be processed
      const element = renderElement as React.ReactElement;
      const cloneProps = React.cloneElement(element).props;
      const cloneChilren = cloneProps.children;

      const childRender: ReactNodeArray = [];
      if (typeof cloneChilren === "string") {
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        processRender(cloneChilren, boldRegex, childRender);
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        processRender(cloneChilren, italicRegex, childRender);
      }

      if (childRender.length) {
        newRender.push(React.cloneElement(element, { children: childRender }));
      } else {
        newRender.push(renderElement);
      }
    }
  });
  render.length = 0;
  render.push(...newRender);
};

const processRender = (
  text: string,
  regex: TagMatchConfig,
  render: ReactNodeArray
) => {
  if (render.length) {
    processNonEmptyRender(regex, render);
  } else {
    processEmptyRender(text, regex, render);
  }
};

export const MarkdownLite = React.memo(({ children }: MarkdownLiteProps) => {
  const inText = useContext(TextContext);
  if (!inText) {
    throw new Error(
      "MarkdownLite components must be rendered within a Text component only."
    );
  }

  const render: ReactNodeArray = [];

  processRender(children, linkRegex, render);
  processRender(children, boldRegex, render);
  processRender(children, italicRegex, render);

  return <>{render.length ? render : children}</>;
});

MarkdownLite.displayName = "MarkdownLite";
