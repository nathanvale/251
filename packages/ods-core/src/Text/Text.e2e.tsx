import React from "react";
import { E2ETests } from "@origin-digital/ods-types";
import { Stack } from "../Stack";
import { Box } from "../Box";
import { Text } from "./Text";

export const tests: E2ETests = [
  {
    label: "Default",
    Code: () => <Text>The quick brown fox jumps over the lazy dog.</Text>,
    responsive: true,
  },
  {
    label: "Body Small Text",
    Code: () => (
      <Text variant="body-small">
        The quick brown fox jumps over the lazy dog.
      </Text>
    ),
    responsive: true,
  },
  {
    label: "Subtitle Text",
    Code: () => (
      <Text variant="subtitle">
        The quick brown fox jumps over the lazy dog.
      </Text>
    ),
  },
  {
    label: "Subtitle Small Text",
    Code: () => (
      <Text variant="subtitle-small">
        The quick brown fox jumps over the lazy dog.
      </Text>
    ),
  },
  {
    label: "Caption Text",
    Code: () => (
      <Text variant="caption">
        The quick brown fox jumps over the lazy dog.
      </Text>
    ),
  },
  {
    label: "Overline Text",
    Code: () => (
      <Text variant="overline-text">
        The quick brown fox jumps over the lazy dog.
      </Text>
    ),
  },
  {
    label: "Tone (positive)",
    Code: () => (
      <Text tone="positive">The quick brown fox jumps over the lazy dog.</Text>
    ),
  },
  {
    label: "Tone (critical)",
    Code: () => (
      <Text tone="critical">The quick brown fox jumps over the lazy dog.</Text>
    ),
  },
  {
    label: "Tone (promoteC)",
    Code: () => (
      <Text tone="promoteC">The quick brown fox jumps over the lazy dog.</Text>
    ),
  },
  {
    label: "Weight (medium)",
    Code: () => (
      <Text weight="medium">The quick brown fox jumps over the lazy dog.</Text>
    ),
  },
  {
    label: "Truncating long text",
    Code: () => (
      <Box style={{ width: 90 }}>
        <Text truncate>The quick brown fox jumps over the lazy dog.</Text>
      </Box>
    ),
  },
  {
    label: "Text Alignment",
    Code: () => (
      <Stack space="medium">
        <Text>The quick brown fox jumps over the lazy dog.</Text>
        <Text align="center">The quick brown fox jumps over the lazy dog.</Text>
        <Text align="right">The quick brown fox jumps over the lazy dog.</Text>
      </Stack>
    ),
  },
  {
    label: "Text Alignment (responsive)",
    Code: () => (
      <Text align={["left", "center"]}>
        The quick brown fox jumps over the lazy dog.
      </Text>
    ),
  },
  {
    label: "Inline Text",
    Code: () => (
      <Text>
        The quick brown fox{" "}
        <Text inline tone="positive">
          jumps over
        </Text>{" "}
        the lazy dog.
      </Text>
    ),
  },
];
