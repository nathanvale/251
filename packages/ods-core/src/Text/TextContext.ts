import { createContext } from "react";
import { UseTextProps } from "./Text";

export const TextContext = createContext<UseTextProps | false>(false);
