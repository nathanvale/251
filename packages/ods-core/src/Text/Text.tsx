import React, { ReactNode, useMemo } from "react";
import clsx from "clsx";
import { BoxProps, Box } from "@origin-digital/ods-core";
import {
  TextVariants,
  TextToneVariants,
  TypographyWeightVariants,
} from "@origin-digital/ods-types";
import {
  useBasekickStyles,
  useToneStyles,
  useWeightStyles,
  useStrongStyles,
} from "@origin-digital/ods-typography";
import {
  useCheckTypographyBackground,
  useTruncatedContent,
} from "../_private/hooks";

import { TextContext } from "./TextContext";

export interface UseTextProps {
  mediumWeight?: boolean;
  variant?: TextVariants;
  tone?: TextToneVariants;
}

export interface TextProps {
  children?: ReactNode;
  variant?: TextVariants;
  tone?: TextToneVariants;
  weight?: Exclude<TypographyWeightVariants, "bold">;
  align?: BoxProps["textAlign"];
  truncate?: boolean;
  "data-id"?: string;
  component?: BoxProps["component"];
}

interface UseTextStylesProps {
  weight: Exclude<TypographyWeightVariants, "bold">;
  tone?: TextToneVariants;
  variant: TextVariants;
}

const defaultWeight = "regular";
const defaultVariant = "body";

function useTextStyles(props: UseTextStylesProps) {
  const { tone = "neutral", weight, variant } = props;
  useCheckTypographyBackground();
  const basekickStyles = useBasekickStyles({ variant });
  const toneStyles = useToneStyles({ tone });
  const weightStyles = useWeightStyles(weight);
  const strongStyles = useStrongStyles("medium");
  return clsx(basekickStyles, toneStyles, weightStyles, strongStyles);
}

export const Text = (props: TextProps) => {
  const {
    children,
    component = "span",
    variant = defaultVariant,
    truncate = false,
    align,
    weight = defaultWeight,
    tone,
    "data-id": dataId,
  } = props;

  // Prevent re-renders when context values haven't changed
  const textContextValue = useMemo(
    () => ({
      tone,
      weight,
      variant,
    }),
    [tone, weight, variant]
  );

  const className = useTextStyles({ weight, tone, variant });

  const content = useTruncatedContent({
    children,
    truncate,
  });

  return (
    <TextContext.Provider value={textContextValue}>
      <Box
        data-id={dataId}
        className={className}
        component={component}
        textAlign={align}
      >
        {content}
      </Box>
    </TextContext.Provider>
  );
};

Text.defaultProps = {
  "data-id": "text",
  component: "span",
  variant: defaultVariant,
  tone: "neutral",
  weight: defaultWeight,
  align: "left",
  truncate: false,
};

Text.displayName = "Text";
