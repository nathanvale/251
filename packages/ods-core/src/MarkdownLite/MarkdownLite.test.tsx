import React from "react";
import {
  render,
  getByTestId as odsGetByTestId,
  getAllByTestId as odsGetAllByTestId,
} from "@origin-digital/ods-testing-library";
import { Text } from "../Text/Text";
import { MarkdownLite } from "./MarkdownLite";

test("It should render bold text using astrix", () => {
  const { container, getByTestId } = render(
    <Text>
      <MarkdownLite>**bold text**</MarkdownLite>
    </Text>
  );

  expect(container.textContent).toBe("bold text");

  const strongText = getByTestId("strong");
  expect(strongText.textContent).toBe("bold text");
});

test("It should render bold text using underscore", () => {
  const { container, getByTestId } = render(
    <Text>
      <MarkdownLite>__bold text__</MarkdownLite>
    </Text>
  );

  expect(container.textContent).toBe("bold text");

  const strongText = getByTestId("strong");
  expect(strongText.textContent).toBe("bold text");
});

test("It should render multiple bold text", () => {
  const { container, getAllByTestId } = render(
    <Text>
      <MarkdownLite>
        this is the **first bold text** and **this is the second** bold text
      </MarkdownLite>
    </Text>
  );

  expect(container.textContent).toBe(
    "this is the first bold text and this is the second bold text"
  );

  const strongText = getAllByTestId("strong");
  expect(strongText[0].textContent).toBe("first bold text");
  expect(strongText[1].textContent).toBe("this is the second");
});

test("It should render italic text using astrix", () => {
  const { container, getByTestId } = render(
    <Text>
      <MarkdownLite>*italic text*</MarkdownLite>
    </Text>
  );

  expect(container.textContent).toBe("italic text");

  const strongText = getByTestId("emphasis");
  expect(strongText.textContent).toBe("italic text");
});

test("It should render italic text using underscore", () => {
  const { container, getByTestId } = render(
    <Text>
      <MarkdownLite>_italic text_</MarkdownLite>
    </Text>
  );

  expect(container.textContent).toBe("italic text");

  const strongText = getByTestId("emphasis");
  expect(strongText.textContent).toBe("italic text");
});

test("It should render multiple italic text", () => {
  const { container, getAllByTestId } = render(
    <Text>
      <MarkdownLite>
        this is the *first italic text* and *this is the second* italic text
      </MarkdownLite>
    </Text>
  );

  expect(container.textContent).toBe(
    "this is the first italic text and this is the second italic text"
  );

  const strongText = getAllByTestId("emphasis");
  expect(strongText[0].textContent).toBe("first italic text");
  expect(strongText[1].textContent).toBe("this is the second");
});

test("should render italic inside strong text", () => {
  const { container, getByTestId } = render(
    <Text>
      <MarkdownLite>
        here is some **bold text with *italic* inside** it
      </MarkdownLite>
    </Text>
  );
  expect(container.textContent).toBe(
    "here is some bold text with italic inside it"
  );

  const bold = getByTestId("strong");
  expect(bold.textContent).toBe("bold text with italic inside");

  const italic = odsGetByTestId(bold, "emphasis");
  expect(italic.textContent).toBe("italic");
});

test("should render a link", () => {
  const { container, getByTestId } = render(
    <Text>
      <MarkdownLite>[Origin Energy](http://origin.com.au)</MarkdownLite>
    </Text>
  );

  expect(container.textContent).toBe("Origin Energy");

  const link = getByTestId("origin-energy");
  expect(link.textContent).toBe("Origin Energy");
  expect(link.getAttribute("href")).toBe("http://origin.com.au");
  expect(link.getAttribute("target")).toBeNull();
});

test("should render two links which are identical", () => {
  const { container, getAllByTestId } = render(
    <Text>
      <MarkdownLite>
        The first [Origin Energy](http://origin.com.au) and the second [Origin
        Energy](http://origin.com.au) should render
      </MarkdownLite>
    </Text>
  );

  expect(container.textContent).toBe(
    "The first Origin Energy and the second Origin Energy should render"
  );

  const links = getAllByTestId("origin-energy");
  expect(links).toHaveLength(2);
  expect(links[0].textContent).toBe("Origin Energy");
  expect(links[0].getAttribute("href")).toBe("http://origin.com.au");
  expect(links[1].textContent).toBe("Origin Energy");
  expect(links[1].getAttribute("href")).toBe("http://origin.com.au");
});

test("should render a link with a target", () => {
  const linkMarkdown =
    '[Origin Energy](http://origin.com.au){:target="_blank"}';
  const { container, getByTestId } = render(
    <Text>
      <MarkdownLite>{linkMarkdown}</MarkdownLite>
    </Text>
  );

  expect(container.textContent).toBe("Origin Energy");

  const link = getByTestId("origin-energy");
  expect(link.textContent).toBe("Origin Energy");
  expect(link.getAttribute("href")).toBe("http://origin.com.au");
  expect(link.getAttribute("target")).toBe("_blank");
});

test("should render a link with bold and italic text within the link", () => {
  const { container, getByTestId } = render(
    <Text>
      <MarkdownLite>
        A link [with **multiple** text that **is bold** and *two blocks* of text
        which are *italic*](http://origin.com.au) as well
      </MarkdownLite>
    </Text>
  );

  expect(container).toMatchSnapshot();
  expect(container.textContent).toBe(
    "A link with multiple text that is bold and two blocks of text which are italic as well"
  );

  const link = getByTestId(
    "with-multiple-text-that-is-bold-and-two-blocks-of-text-which-are-italic"
  );
  expect(link.textContent).toBe(
    "with multiple text that is bold and two blocks of text which are italic"
  );
  expect(link.getAttribute("href")).toBe("http://origin.com.au");

  const bold = odsGetAllByTestId(link, "strong");
  expect(bold[0].textContent).toBe("multiple");
  expect(bold[1].textContent).toBe("is bold");

  const italic = odsGetAllByTestId(link, "emphasis");
  expect(italic[0].textContent).toBe("two blocks");
  expect(italic[1].textContent).toBe("italic");
});

[
  "This is **supposed to be bold but is not closed",
  "This is *supposed to be italic but is not closed",
  "This is **supposed to be bold*??",
].forEach((md) => {
  test(`should return '${md}'`, () => {
    const { container } = render(
      <Text>
        <MarkdownLite>{md}</MarkdownLite>
      </Text>
    );
    expect(container.textContent).toBe(md);
  });
});

test("should return text with italics", () => {
  const { container, getByTestId } = render(
    <Text>
      <MarkdownLite>This is *supposed to be bold**??</MarkdownLite>
    </Text>
  );

  expect(container.textContent).toBe("This is supposed to be bold*??");

  const italic = getByTestId("emphasis");
  expect(italic.textContent).toBe("supposed to be bold");
});

test("should throw an error if not wrapped in a Text component", () => {
  expect(() => render(<MarkdownLite>*bold text*</MarkdownLite>)).toThrow();
});
