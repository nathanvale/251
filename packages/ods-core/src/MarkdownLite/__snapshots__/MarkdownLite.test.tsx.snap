// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a link with bold and italic text within the link 1`] = `
.c0 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  max-width: 100%;
  text-align: left;
}

<div>
  <span
    class="typography-body-variant-basekick-64 typography-neutral-color-65 typography-regular-weight-66 typography-strong-support-67 -span c0"
    data-id="text"
  >
    A link 
    <a
      class="typography-linkReset-69 typography-neutral-tone-link-70"
      data-id="with-multiple-text-that-is-bold-and-two-blocks-of-text-which-are-italic"
      href="http://origin.com.au"
    >
      with 
      <strong
        class="typography-medium-weight-71"
        data-id="strong"
      >
        multiple
      </strong>
       text that 
      <strong
        class="typography-medium-weight-72"
        data-id="strong"
      >
        is bold
      </strong>
       and 
      <em
        data-id="emphasis"
      >
        two blocks
      </em>
       of text which are 
      <em
        data-id="emphasis"
      >
        italic
      </em>
      
    </a>
     as well
  </span>
</div>
`;
